### DISCUSSION TITLE
User supplied sentiment matching dictionary?

### DISCUSSION THREAD

**Discussion by ChRauh** [BenWiseman/sentiment.ai#6] [Q&A] [answered]
Dear Ben,

Many thanks for sharing this tremendous work (and especially for the hassle-free setup and communication with the Python environment)!

I was wondering whether I can supply/call my own sentiment matching dictionary somewhere instead of what is in data(default). 
I understand the neat sentiment_match() function, but I am not interested in the closest matching "phrase" but rather in the (average) similarity to the whole category. But as far as I can see, the sentiment score returned by sentiment_match() is identical to the one sentiment_score() produces, right?

Any hint would be appreciated and sorry if I am missing something.

Best,
Christian

**Comment by BenWiseman** [comment, marked as answer]
You are correct that the sentiment in sentiment_match() is the same as sentiment_score() - this is deliberate for consistency such that _match() simply gives more information. 

You can supply your own phrases, e.g.
```
my_comments <- c("Happy labradoor fetching a ball", 
                                "Playful golden retriever at the park",
                                "Angry cat getting a belly rub",
                                "Sweet cat sniffing catnip")

dogs <- c("excited dog", "playful dog", "happy dog", "unhappy dog")
cats <- c("excited cat", "playful cat", "happy cat", "unhappy cat")

my_categories <- list(dogs = dogs, cats = cats)

result <- sentiment_match(x = my_comments, phrases = my_categories)
```

Now, for average similarity to the whole category, it's not a built in function, so you'd need to use our included matrix helpers, and a bit of data.table (or dplyr if you swing that way) magic like so:

```
comm_emb   <- embed_text(my_comments)
dogs_emb   <- embed_text(dogs)
cats_emb   <- embed_text(cats)
all_emb    <- rbind(dogs_emb, cats_emb)
result_raw <- cosine_match(comm_emb, all_emb)

# avg similarity
# create table of categories
ref_dt <- data.table(phrase = c(dogs, cats), 
                     topic   = c(rep("dogs", length(dogs)),
                                 rep("cats", length(cats))))
# join
result_raw <- ref_dt[result_raw, on = c(phrase = "reference")]

# summarise and viola!
result_raw[, .(avg_similarity = mean(similarity)), by = c("target", "topic")]
```
If you wanted average sentiment to all phrases, you could add a column of `sentiment_score()` values to `ref_dt` before joining it and do `weighted_mean()` of sentiment using similarity as a weight vector.

**Reply by ChRauh** [reply]
Great! Exactly what I was looking for and learned a bit of data.table on the way. Many thanks for your work and support!
